src/v/utils/tracking_allocator.h

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "utils/intrusive_list_helpers.h"
+
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/sstring.hh>
+
+#include <memory>
+
+namespace ss = seastar;
+
+namespace util {
+
+/// Tracks allocations and deallocations of memory. A mem_tracker can have
+/// child trackers effectively forming an n-ary tree like structure.
+class mem_tracker {
+public:
+    explicit mem_tracker(ss::sstring label)
+      : _label(std::move(label)) {}
+    mem_tracker(mem_tracker&) = delete;
+    mem_tracker(mem_tracker&&) = default;
+    mem_tracker& operator=(mem_tracker&) = delete;
+    mem_tracker& operator=(mem_tracker&&) = delete;
+    ~mem_tracker() = default;
+
+    void allocate(int64_t bytes) { _consumption += bytes; }
+
+    void deallocate(int64_t bytes) { _consumption -= bytes; }
+
+    ss::lw_shared_ptr<mem_tracker> create_child(ss::sstring);
+
+    /// Consumption of self + all the children rooted with this tracker..
+    int64_t consumption() const;
+
+    /// Pretty prints the tracker tree rooted at this instance.
+    /// Example:
+    /// |
+    /// root: 32.000bytes
+    /// | |
+    /// | child1: 24.000bytes
+    /// | | |
+    /// | | child11: 0.000bytes
+    /// | | | |
+    /// | | | child111: 0.000bytes
+    /// | |
+    /// | child2: 0.000bytes
+    ss::sstring debug(int level = 0) const;
+
+private:
+    ss::sstring _label;
+    int64_t _consumption = 0;
+    intrusive_list_hook _hook;
+    intrusive_list<mem_tracker, &mem_tracker::_hook> _children;
+};
+
+/// An allocator paired with a memory tracker to account for consumption.
+template<class T, class allocator = std::allocator<T>>
+class tracking_allocator : public allocator {
```

Comment:
Is this the right way to extend an underlying allocator, in this case `allocator`?

Allocators provide many types & methods to the outer container, via `allocator_traits` but here we don't forward any of them from the underlying allocator, so you will end up with default values. It seems like to do this right you should forward the all of the things inspected by [allocator_traits](https://en.cppreference.com/w/cpp/memory/allocator_traits) to the underlying allocator (using `allocator_traits<allocator>::thing`), except those that are changed by the addition of tracking (e.g., `is_always_equal` should be always false regardless of the underlying allocator).